resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -565,7 +565,6 @@ private[spark] class Client(
           distribute(jarsArchive.toURI.getPath,
             resType = LocalResourceType.ARCHIVE,
             destName = Some(LOCALIZED_LIB_DIR))
-          jarsArchive.delete()
```

Comment:
You're undoing the fix for SPARK-20741. If this is causing a problem and you want to fix it, you need to make it so that you don't do this only when the specific scenario that's causing the problem happens.

Reply:
Agree with Marcelo, this is a valid concern, we should not avoid such regression here.

Reply:
Thanks @vanzin for the pointer. It was my mistake, I missed the change reason while looking at the history of the file.

I still see that SPARK-20741 has fixed the issue partially, it leaves \_\_spark_conf\_\_*.zip file to delete as part of shutdownhook.

I see these approaches to fix it further,

1. Delete \_\_spark_conf\_\_*.zip and \_\_spark_libs\_\_*.zip files after completing the application similar to cleanupStagingDir. 
(Or)
2. Add a configuration whether to delete \_\_spark_conf\_\_*.zip and \_\_spark_libs\_\_*.zip files after copying to dest dir, so that users can decide whether to delete these immediately or as part of process exit. In case of SPARK-20741, this new configuration can be enabled to delete these files immediately. 

@vanzin & @jerryshao Please let me know your thoughts on this or if you have any other way to do this. Thanks

Reply:
What if your scenario and SPARK-20741's scenario are both encountered? Looks like your approach above cannot be worked.

I'm wondering if we can copy or move this __spark_libs__.zip temp file to another non-temp file and add that file to the dist cache. That non-temp file will not be deleted and can be overwritten during another launching, so we will always have only one copy.

Besides, I think we have several workarounds to handle this issue like spark.yarn.jars or spark.yarn.archive, so looks like this corner case is not so necessary to fix (just my thinking, normally people will not use local FS in a real cluster).

Reply:
Thanks @jerryshao for the comment.

> What if your scenario and SPARK-20741's scenario are both encountered? Looks like your approach above cannot be worked.

Can you provide some information why you think it doesn't work? If we delete the spark_libs.zip after completing the application(similar to staging dir deletion), it would not stack up till the process exit which solves SPARK-20741 and also becomes available during the execution for this current issue. 
> I'm wondering if we can copy or move this spark_libs.zip temp file to another non-temp file and add that file to the dist cache. That non-temp file will not be deleted and can be overwritten during another launching, so we will always have only one copy.

If there are multiple jobs submitted/running concurrently, we would be overwriting the existing with the latest spark_libs.zip which may lead to apps failure during the copy-in-progress and also would become ambiguous to delete the file by which application.

> Besides, I think we have several workarounds to handle this issue like spark.yarn.jars or spark.yarn.archive, so looks like this corner case is not so necessary to fix (just my thinking, normally people will not use local FS in a real cluster).

I agree, this is a corner case and can be handled with workaround.

Reply:
Think about this again, I think you're right. But I'm not sure if the program will be crashed or not if we delete the dependencies in the run-time.

Reply:
Adding a configuration is rarely the right fix for anything. You shouldn't have to choose to have the correct behavior or not.

You can probably fix this by always uploading things to the distributed cache for resources with a "file:" scheme. This will penalize those running things with a local default FS, but it will work. And the 99.9% of the world that does do that will not be affected.

Reply:
Thanks @vanzin for the comment, I will update the PR as per your suggestion.